File: leetcode-04/solutions/python3/0016.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        if len(nums) < 3:
            return 0

        ans = nums[0] + nums[1] + nums[2]

        nums.sort()

        for i in range(len(nums) - 2):
            l = i + 1
            r = len(nums) - 1
            while l < r:
                sum = nums[i] + nums[l] + nums[r]
                if sum == target:
                    return sum
                elif abs(sum - target) < abs(ans - target):
                    ans = sum
                elif sum < target:
                    l += 1
                else:
                    r -= 1

        return ans

File: leetcode-04/solutions/cpp/0016.cpp
class Solution {
 public:
  int threeSumClosest(vector<int>& nums, int target) {
    if (nums.size() < 3) return 0;

    int ans = nums[0] + nums[1] + nums[2];

    sort(nums.begin(), nums.end());

    for (int i = 0; i < nums.size() - 2; i++) {
      int l = i + 1;
      int r = nums.size() - 1;
      while (l < r) {
        int sum = nums[i] + nums[l] + nums[r];
        if (sum == target)
          return sum;
        else if (abs(sum - target) < abs(ans - target))
          ans = sum;
        else if (sum < target)
          l++;
        else
          r--;
      }
    }

    return ans;
  }
};